### PR TITLE
ZCS-12912: LDAP global attributes latest value not available across node

### PR DESCRIPTION
**Pull request by swaatiTelus**
Problem ([ZCS-12912](https://jira.corp.synacor.com/browse/ZCS-12912)):
When we add or delete external S3 volume on one server, the LDAP global attribute `zimbraGlobalExternalStoreConfig` value is not updated across other mail servers. So on other servers, we have to execute `zmprov fc all`, to get the updated value of LDAP reflected.

Solution:
The cache should be flushed at _config level_ at the time of creating and deleting the S3 Volume. Cache should be  flushed for all the servers other than the server where changes were made, so that the updated LDAP value is reflected among all the servers right away.

Fix:
Executing a flushCache method after each S3 external volume addition and deletion.

Testing:
Done testing on multi-node server after adding and deleting multiple volumes.

Other PR: https://github.com/Zimbra/zm-hsm-store/pull/42

### CODE REVIEW

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 1]
```diff
@@ -157,13 +164,52 @@ public void addServerProperties(VolumeInfo volInfo, String serverId) throws Serv
...
+     */
+    private void flushCache(ZimbraSoapContext zsc) {
+        //In this option we have the dependency on zcs parameter.
+        try {
+          CacheSelector cacheSelector = new CacheSelector();
+          cacheSelector.setTypes(CacheEntryType.config.name());
+          FlushCache.flushConfigCacheOnAllServers(zsc, new FlushCacheRequest(cacheSelector));
+        } catch (ServiceException e) {
+          ZimbraLog.account.info("Encountered exception during FlushCache after creating unified volume" + e.getMessage());
+          e.printStackTrace();
```
Comment (RakeshAMore) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: Use logger instead

**Comment by RakeshAMore** [comment]
As discussed please make changes accordingly.

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 2]
```diff
@@ -271,9 +317,11 @@ public void editGlobalConfigOnDeleteVolume(int volumeId, String serverId) throws
                    globalExtStoreConfJson.put(GlobalExternalStoreConfigConstants.A_S3_UNIFIED_VOLUME, unifiedJsonObject);
                    Provisioning.getInstance().getConfig().setGlobalExternalStoreConfig(globalExtStoreConfJson.toString());
                 }
+                //flushCache(zsc);
```
Comment (abhishekdeshmukh2209) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: commented code

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 3]
```diff
@@ -157,13 +164,52 @@ public void addServerProperties(VolumeInfo volInfo, String serverId) throws Serv
                     && AdminConstants.A_VOLUME_S3.equalsIgnoreCase(volInfo.getVolumeExternalInfo().getStorageType())
                     && volInfo.getVolumeExternalInfo().isUnified()) {
                 editGlobalConfigOnAddVolume(volInfo, serverId);
+                //flushCache(zsc);
```
Comment (abhishekdeshmukh2209) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: commented code

File: store/src/java/com/zimbra/cs/service/admin/FlushCache.java [hunk 4]
```diff
@@ -328,4 +328,10 @@ private static String sanitizeImapCacheTypes(String cacheTypes) {
         }
         return imapTypes.isEmpty() ? null : Joiner.on(",").join(imapTypes);
     }
+
```
Comment (abhishekdeshmukh2209) [inline store/src/java/com/zimbra/cs/service/admin/FlushCache.java]: extra lines

Comment (swaatiTelus) [reply store/src/java/com/zimbra/cs/service/admin/FlushCache.java]: removed.

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java (hunk 2 above)
Comment (swaatiTelus) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: The approach was under discussion in the draft PR. Uncommented it. Thanks!

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java (hunk 3 above)
Comment (swaatiTelus) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: Uncommented it.

**Review by ronstra-synacor — APPROVED** [review]

File: store/src/java/com/zimbra/cs/service/admin/FlushCache.java [hunk 5]
```diff
@@ -328,4 +328,9 @@ private static String sanitizeImapCacheTypes(String cacheTypes) {
...
+    public static void flushConfigCacheOnAllServers(ZimbraSoapContext zsc, FlushCacheRequest req) throws ServiceException {
```
Comment (RakeshAMore) [inline store/src/java/com/zimbra/cs/service/admin/FlushCache.java]: Since it's named to flush the cache of the config only then how will it prevent another type of cache flush?

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 6]
```diff
@@ -157,13 +163,28 @@ public void addServerProperties(VolumeInfo volInfo, String serverId) throws Serv
...
             throw e;
         }
     }
 
+    /**
+     * Flushes config level cache of all other server
+     * @param zsc
+     * @return nothing
+     */
+    public void flushCache(ZimbraSoapContext zsc) {
```
Comment (RakeshAMore) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: I would name it to specific attributes

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 7]
```diff
@@ -157,13 +163,28 @@ public void addServerProperties(VolumeInfo volInfo, String serverId) throws Serv
...
+        try {
+          CacheSelector cacheSelector = new CacheSelector();
+          cacheSelector.setTypes(CacheEntryType.config.toString());
+          FlushCache.flushConfigCacheOnAllServers(zsc, new FlushCacheRequest(cacheSelector));
+        } catch (ServiceException e) {
+          ZimbraLog.account.info("Encountered exception during FlushCache after creating unified volume " + e.getMessage());
```
Comment (RakeshAMore) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: ZimbraLog.account.info("Encountered exception during FlushCache after creating unified volume ", e);

Comment (swaatiTelus) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: done

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java (hunk 6 above)
Comment (swaatiTelus) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: Renamed method name to flushConfigLevelCacheOnAllServers().

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java (hunk 1 above)
Comment (swaatiTelus) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: Done.

File: store/src/java/com/zimbra/cs/service/admin/FlushCache.java (hunk 5 above)
Comment (swaatiTelus) [reply store/src/java/com/zimbra/cs/service/admin/FlushCache.java]: Yes right, so now I have moved the logic of defining the cache type inside this method itself. Thanks!

File: store/src/java/com/zimbra/cs/service/admin/FlushCache.java [hunk 8]
```diff
@@ -328,4 +328,16 @@ private static String sanitizeImapCacheTypes(String cacheTypes) {
...
         return imapTypes.isEmpty() ? null : Joiner.on(",").join(imapTypes);
     }
+
+    /**
+     * Flushes config level cache of all other server
+     * @param zimbraSoapContext
+     * @return nothing
+     * @throws ServiceException
+     */
+    public static void flushConfigCacheOnAllServers(ZimbraSoapContext zimbraSoapContext) throws ServiceException {
```
Comment (silentsakky) [inline store/src/java/com/zimbra/cs/service/admin/FlushCache.java]: not sure if we should start adding different methods based on type of cache which needs to be flushed
any caller can do this to flush cache on all servers
```
                CacheSelector cacheSel = new CacheSelector(true /* allServers */, CacheEntryType.config.toString());
                FlushCacheRequest fcReq = new FlushCacheRequest(cacheSel);
                try {
                    FlushCache.doFlushCache(this, context, fcReq);
                } catch (ServiceException se) {
                    // Error log
                }
```

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 9]
```diff
@@ -102,7 +107,10 @@ public void deleteServerProperties(int volumeId, String serverId) throws Service
             // step 6: Set ldap attribute
             provisioning.getLocalServer().setServerExternalStoreConfig(currentJsonObject.toString());
 
-            editGlobalConfigOnDeleteVolume(volumeId, serverId);
+            //if volume is unified, then only we should call the below method to modify global external store
+            if(isUnified) {
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]:
```suggestion
            if (isUnified) {
```

File: store/src/java/com/zimbra/cs/service/admin/FlushCache.java (hunk 8 above)
Comment (RakeshAMore) [reply store/src/java/com/zimbra/cs/service/admin/FlushCache.java]: It is better to restrict it to config only because flushing another type of cache on all servers may have an impact on the performance so it's better to not add if this is not required :)(and if we have to add then impact should be analyzed before.)

Comment (silentsakky) [reply store/src/java/com/zimbra/cs/service/admin/FlushCache.java]: if you check my code then it's restricting to config cache only :)

Comment (RakeshAMore) [reply store/src/java/com/zimbra/cs/service/admin/FlushCache.java]: Got it. Swati will make changes accordingly. Thanks

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 10]
```diff
@@ -102,15 +113,19 @@ public void deleteServerProperties(int volumeId, String serverId) throws Service
             // step 6: Set ldap attribute
             provisioning.getLocalServer().setServerExternalStoreConfig(currentJsonObject.toString());
 
-            editGlobalConfigOnDeleteVolume(volumeId, serverId);
-
+            //if volume is unified, then only we should call the below method to modify global external store
+            if (isUnifiedVolume) {
+                editGlobalConfigOnDeleteVolume(volumeId, serverId);
```
Comment (silentsakky) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:115]: can't we call flushConfigLevelCacheOnAllServers from here itself? we are changing function signatures to return isUnifiedVolume flag and at the end based on that flag calling flushConfigLevelCacheOnAllServers, which can be done here also

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 11]
```diff
@@ -360,4 +376,18 @@ public Boolean isVolumePresentInJson(int volumeId) throws ServiceException, JSON
...
+     * Performs flush cache at config level on all the server
+     * @param adminDocumentHandler, context
+     * @return nothing
+     */
+    public static void flushConfigLevelCacheOnAllServers(AdminDocumentHandler adminDocumentHandler, Map<String, Object> context) {
+        CacheSelector cacheSelector = new CacheSelector(true, CacheEntryType.config.toString());
+        try {
+            FlushCache.doFlushCache(adminDocumentHandler, context, new FlushCacheRequest(cacheSelector));
+        } catch (ServiceException se) {
+            ZimbraLog.account.error("Encountered exception during FlushCache: ", se);
```
Comment (silentsakky) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: not sure we should log to account category as this operation is not related to any account

Comment (swaatiTelus) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: Done. Changed it to `misc` as used in FlushCache utility.

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java (hunk 10 above)
Comment (swaatiTelus) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:115]: Yes right, we could have done it by 2 ways: 
1. Carry the AdminDocumentHandler reference as a parameter from DeleteVolume API up to here to handler class. and
2. to send back the boolean value from handler.
I opted for option 2, to maintain the uniformity with other API, as we are calling `flushConfigLevelCacheOnAllServers` method from API's itself.

Please suggest if I should change it.?

Comment (silentsakky) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:115]: ok I missed the handler part, option 1 is ok

**Review by silentsakky — APPROVED** [review]

**Review by silentsakky — APPROVED** [review]

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 12]
```diff
@@ -360,4 +370,41 @@ public Boolean isVolumePresentInJson(int volumeId) throws ServiceException, JSON
...
+    /**
+     * Checks if volume is unified or not
+     * @param volumeId
+     * @return true if volume is unified else false
+     */
+    public boolean isUnifiedVolume(int volumeId){
+        boolean isUnifiedVolume = false;
+        try {
+            JSONObject currentJsonObject = new JSONObject(provisioning.getLocalServer().getServerExternalStoreConfig());
+            JSONArray serverStoreJsonArray = currentJsonObject.getJSONArray("server/stores");
```
Comment (vishnukantjangid) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:397]: We should use constant for the key "server/stores"

**Review by zimsuchitgupta — APPROVED** [review]

**Review by vishnukantjangid — APPROVED** [review]
